### ISSUE TITLE
Comparison with  your  earlier paper

### ISSUE THREAD

**Issue by 009deep** [clovaai/voxceleb_trainer#10]
I need some clarity on current implementation.

1. There is difference in thin resnet mentioned in your [earlier work](http://www.robots.ox.ac.uk/~vgg/publications/2019/Nagrani19/nagrani19.pdf) vs  what is implemented here. From number of blocks in each layers to number of filters in different blocks and layers. Resulting in model size to reduce to great extent. Parameters count reduced from over 8M parameter to less than 1.5 M.  Is there   any paper that details thin resntet, how do you go about selecting number of blocks and number of filters in each block?

2. In earlier paper you used VLAD and  GHOST layers which are removed here. Do you find them unnecessary with current implementation?

3. This one is not related to old paper but I am bit confused at the moment with current implementation. TEER   means different for different type of losses. In case of angleproto loss, when you get  optimized result (least value  of VEER) as shown paper, what value of TEER (which I think is accuracy) do you get? 
I am training model using only voxceleb1 training data and I am getting following result at 400 epochs:

    > TEER 70.15, TLOSS 1.085345, VEER 1.8081

    I am surprised with this result as VEER here is already less than what is mentioned in paper, so something doesn't seem quite right here. Any suggestion?

**Comment by joonson** [comment]
1. The Thin ResNet (`ResNet34L`) implemented was used in this paper.
```
@inproceedings{chung2019delving,
  title={Delving into VoxCeleb: environment invariant speaker recognition},
  author={Chung, Joon Son and Huh, Jaesung and Mun, Seongkyu},
  booktitle={Speaker Odyssey},
  year={2020}
}
```

2. They show similar performance to the SAP layers in the model.

3. TEER can mean EER or accuracy depending on the loss function. VEER seems very low for a VoxCeleb1 trained model. Can you send me the train and test lists by email so I can try to reproduce the results?

**Comment by 009deep** [comment]
Thanks  for response. I'll e-mail other details.

**Comment by joonson** [comment]
I checked the email and this is due to train-test contamination. Identities 10270 to 10309 are reserved for testing, but they are also in your training set.

**Comment by jlian2** [comment]
> I need some clarity on current implementation.
> 
> 1. There is difference in thin resnet mentioned in your [earlier work](http://www.robots.ox.ac.uk/~vgg/publications/2019/Nagrani19/nagrani19.pdf) vs  what is implemented here. From number of blocks in each layers to number of filters in different blocks and layers. Resulting in model size to reduce to great extent. Parameters count reduced from over 8M parameter to less than 1.5 M.  Is there   any paper that details thin resntet, how do you go about selecting number of blocks and number of filters in each block?
> 2. In earlier paper you used VLAD and  GHOST layers which are removed here. Do you find them unnecessary with current implementation?
> 3. This one is not related to old paper but I am bit confused at the moment with current implementation. TEER   means different for different type of losses. In case of angleproto loss, when you get  optimized result (least value  of VEER) as shown paper, what value of TEER (which I think is accuracy) do you get?
>    I am training model using only voxceleb1 training data and I am getting following result at 400 epochs:
>    > TEER 70.15, TLOSS 1.085345, VEER 1.8081
>    
>    
>    I am surprised with this result as VEER here is already less than what is mentioned in paper, so something doesn't seem quite right here. Any suggestion?

Really interested in the discussions. May I know TEER means training EER? while VEER means valication EER?

**Comment by jlian2** [comment]
I meet the same issue. Have you figured out why TEER is so high?

**Comment by joonson** [comment]
TEER refers to train accuracy in proto, angleproto, ge2e and softmax-based losses, and EER in pairwise-based losses. I will update this in the next commit to avoid confusion. VEER always refers to validation EER.

**Comment by 009deep** [comment]
Thanks for catching that train/test mix up part. I missed it. 

Coming back to TEER part, I understand that in non softmax loss function it represent accuracy. Still have couple of questions:

1.  I am bit confused with following implementation in angleproto. I tried to understand from paper 
 but still not clear. How does it  work? Labels values are just integer ranged from 0 to stepsize? Also what is dimension of input x here? I tried to debug with using  (batchsize, nOut) and (batchsize, 1, nOut) both fails.

```
        out_anchor      = torch.mean(x[:,1:,:],1)
        out_positive    = x[:,0,:]
        stepsize        = out_anchor.size()[0]
        cos_sim_matrix  = F.cosine_similarity(out_positive.unsqueeze(-1).expand(-1,-1,stepsize),out_anchor.unsqueeze(-1).expand(-1,-1,stepsize).transpose(0,2))
        torch.clamp(self.w, 1e-6)
        cos_sim_matrix = cos_sim_matrix * self.w + self.b        
        label       = torch.from_numpy(numpy.asarray(range(0,stepsize))).cuda() 
```

2. Even though TEER represent accuracy, do you have value of TEER for angleproto for result presented in paper. It's just as reference to get sense of what is comparable value of TEER.

**Comment by joonson** [comment]
1. The input to the loss function should be (batchSize, M, nOut) where M>=2. The positive pairs are along the diagonal in the cosine similarity matrix, hence the GT labels are `range(0,stepsize))`.
2. TEER here is the N-way matching accuracy. With these training parameters `--model=ResNetSE34L --trainfunc proto --nSpeakers 2 --batch_size=400 --max_seg_per_spk=100 --max_frames=200` we get TEER of 72% after 500 epochs.

**Comment by 009deep** [comment]
Thanks Joon Son. This details are useful. Will it be feasible, to get same details for other losses, so I I can use them as reference for my own training.

**Comment by joonson** [comment]
What details do you need on the other losses?

**Comment by 009deep** [comment]
If you can e-mail me scores.txt for each losses that would be great. Or if you can add in comment here similar to above one for each loss. e.g

> TEER here is the N-way matching accuracy. With these training parameters --model=ResNetSE34L --trainfunc proto --nSpeakers 2 --batch_size=400 --max_seg_per_spk=100 --max_frames=200 we get TEER of 72% after 500 epochs.

**Comment by joonson** [comment]
Here are a few samples:
[ge2e3.txt](https://github.com/clovaai/voxceleb_trainer/files/4533323/ge2e3.txt)
[amsoft_30_02.txt](https://github.com/clovaai/voxceleb_trainer/files/4533324/amsoft_30_02.txt)
[angleproto2.txt](https://github.com/clovaai/voxceleb_trainer/files/4533327/angleproto2.txt)
[proto2.txt](https://github.com/clovaai/voxceleb_trainer/files/4533328/proto2.txt)

**Comment by 009deep** [comment]
thanks a lot
